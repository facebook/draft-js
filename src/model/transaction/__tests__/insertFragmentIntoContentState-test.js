/**
 * Copyright (c) 2013-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @emails oncall+ui_infra
 */

'use strict';

jest.disableAutomock();

const insertFragmentIntoContentState = require('insertFragmentIntoContentState');

const getSampleStateForTesting = require('getSampleStateForTesting');
const getSampleStateForTestingNestedBlocks = require('getSampleStateForTestingNestedBlocks');

describe('insertFragmentIntoContentState+', () => {
  it('must randomize fragment keys and insert blocks at the end of last block text', () => {
    const {
      contentState,
      selectionState,
    } = getSampleStateForTesting();

    const blockMap = contentState.getBlockMap();
    const blockKeys = blockMap.keySeq().toArray();

    const fisrtBlock = blockMap.first();
    const lastBlock = blockMap.last();

    const selection = selectionState.merge({
      focusKey: lastBlock.getKey(),
      anchorKey: lastBlock.getKey(),
      focusOffset: lastBlock.getLength(),
      anchorOffset:  lastBlock.getLength()
    });

    // we are trying to use the current blockMap to insert and replace the existing one
    const fragmentBlockMap = contentState.getBlockMap();
    const newKeys = fragmentBlockMap.keySeq().toArray();

    const newContentState = insertFragmentIntoContentState(
      contentState,
      selection,
      fragmentBlockMap
    );

    const newBlockMap = newContentState.getBlockMap();

    expect(blockKeys).not.toBe(newKeys);
    expect(lastBlock.getText() + fisrtBlock.getText()).toBe(newBlockMap.skip(2).first().getText());
    expect(blockMap.last().getText()).toBe(newBlockMap.last().getText());
  });

  it('must randomize fragment keys and insert blocks at the end of last block text with nesting enabled', () => {
    const {
      contentState,
      selectionState,
    } = getSampleStateForTestingNestedBlocks();

    const blockMap = contentState.getBlockMap();
    const blockKeys = blockMap.keySeq().toArray();

    const fisrtBlock = blockMap.first();
    const lastBlock = blockMap.last();

    const selection = selectionState.merge({
      focusKey: lastBlock.getKey(),
      anchorKey: lastBlock.getKey(),
      focusOffset: lastBlock.getLength(),
      anchorOffset:  lastBlock.getLength()
    });

    // we are trying to use the current blockMap to insert and replace the existing one
    const fragmentBlockMap = contentState.getBlockMap();
    const newKeys = fragmentBlockMap.keySeq().toArray();

    const newContentState = insertFragmentIntoContentState(
      contentState,
      selection,
      fragmentBlockMap
    );

    const newBlockMap = newContentState.getBlockMap();

    expect(blockKeys).not.toBe(newKeys);
    expect(lastBlock.getText() + fisrtBlock.getText()).toBe(newBlockMap.skip(5).first().getText());
    expect(blockMap.last().getText()).toBe(newBlockMap.last().getText());
  });
});
