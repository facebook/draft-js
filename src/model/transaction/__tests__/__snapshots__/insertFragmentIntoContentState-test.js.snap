// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`must apply fragment at the end 1`] = `
Immutable.Record {
  "entityMap": Immutable.OrderedMap {},
  "blockMap": Immutable.OrderedMap {
    "a": Immutable.Record {
      "key": "a",
      "type": "unstyled",
      "text": "Alphaxx",
      "characterList": Immutable.List [
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
      ],
      "depth": 0,
      "data": Immutable.Map {
        "a": 1,
      },
    },
    "b": Immutable.Record {
      "key": "b",
      "type": "unordered-list-item",
      "text": "Bravo",
      "characterList": Immutable.List [
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "BOLD",
          ],
          "entity": "1",
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "BOLD",
          ],
          "entity": "1",
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "BOLD",
          ],
          "entity": "1",
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "BOLD",
          ],
          "entity": "1",
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "BOLD",
          ],
          "entity": "1",
        },
      ],
      "depth": 0,
      "data": Immutable.Map {},
    },
    "c": Immutable.Record {
      "key": "c",
      "type": "code-block",
      "text": "Test",
      "characterList": Immutable.List [
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
      ],
      "depth": 0,
      "data": Immutable.Map {},
    },
    "d": Immutable.Record {
      "key": "d",
      "type": "code-block",
      "text": "",
      "characterList": Immutable.List [],
      "depth": 0,
      "data": Immutable.Map {},
    },
    "e": Immutable.Record {
      "key": "e",
      "type": "code-block",
      "text": "",
      "characterList": Immutable.List [],
      "depth": 0,
      "data": Immutable.Map {},
    },
    "f": Immutable.Record {
      "key": "f",
      "type": "blockquote",
      "text": "Charlie",
      "characterList": Immutable.List [
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "ITALIC",
          ],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "ITALIC",
          ],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "ITALIC",
          ],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "ITALIC",
          ],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "ITALIC",
          ],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "ITALIC",
          ],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "ITALIC",
          ],
          "entity": null,
        },
      ],
      "depth": 0,
      "data": Immutable.Map {},
    },
  },
  "selectionBefore": Immutable.Record {
    "anchorKey": "a",
    "anchorOffset": 5,
    "focusKey": "a",
    "focusOffset": 5,
    "isBackward": false,
    "hasFocus": true,
  },
  "selectionAfter": Immutable.Record {
    "anchorKey": "a",
    "anchorOffset": 7,
    "focusKey": "a",
    "focusOffset": 7,
    "isBackward": false,
    "hasFocus": true,
  },
}
`;

exports[`must apply fragment to the start 1`] = `
Immutable.Record {
  "entityMap": Immutable.OrderedMap {},
  "blockMap": Immutable.OrderedMap {
    "a": Immutable.Record {
      "key": "a",
      "type": "unstyled",
      "text": "xxAlpha",
      "characterList": Immutable.List [
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
      ],
      "depth": 0,
      "data": Immutable.Map {
        "a": 1,
      },
    },
    "b": Immutable.Record {
      "key": "b",
      "type": "unordered-list-item",
      "text": "Bravo",
      "characterList": Immutable.List [
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "BOLD",
          ],
          "entity": "1",
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "BOLD",
          ],
          "entity": "1",
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "BOLD",
          ],
          "entity": "1",
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "BOLD",
          ],
          "entity": "1",
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "BOLD",
          ],
          "entity": "1",
        },
      ],
      "depth": 0,
      "data": Immutable.Map {},
    },
    "c": Immutable.Record {
      "key": "c",
      "type": "code-block",
      "text": "Test",
      "characterList": Immutable.List [
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
      ],
      "depth": 0,
      "data": Immutable.Map {},
    },
    "d": Immutable.Record {
      "key": "d",
      "type": "code-block",
      "text": "",
      "characterList": Immutable.List [],
      "depth": 0,
      "data": Immutable.Map {},
    },
    "e": Immutable.Record {
      "key": "e",
      "type": "code-block",
      "text": "",
      "characterList": Immutable.List [],
      "depth": 0,
      "data": Immutable.Map {},
    },
    "f": Immutable.Record {
      "key": "f",
      "type": "blockquote",
      "text": "Charlie",
      "characterList": Immutable.List [
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "ITALIC",
          ],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "ITALIC",
          ],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "ITALIC",
          ],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "ITALIC",
          ],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "ITALIC",
          ],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "ITALIC",
          ],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "ITALIC",
          ],
          "entity": null,
        },
      ],
      "depth": 0,
      "data": Immutable.Map {},
    },
  },
  "selectionBefore": Immutable.Record {
    "anchorKey": "a",
    "anchorOffset": 0,
    "focusKey": "a",
    "focusOffset": 0,
    "isBackward": false,
    "hasFocus": true,
  },
  "selectionAfter": Immutable.Record {
    "anchorKey": "a",
    "anchorOffset": 2,
    "focusKey": "a",
    "focusOffset": 2,
    "isBackward": false,
    "hasFocus": true,
  },
}
`;

exports[`must apply fragment to within block 1`] = `
Immutable.Record {
  "entityMap": Immutable.OrderedMap {},
  "blockMap": Immutable.OrderedMap {
    "a": Immutable.Record {
      "key": "a",
      "type": "unstyled",
      "text": "Alxxpha",
      "characterList": Immutable.List [
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
      ],
      "depth": 0,
      "data": Immutable.Map {
        "a": 1,
      },
    },
    "b": Immutable.Record {
      "key": "b",
      "type": "unordered-list-item",
      "text": "Bravo",
      "characterList": Immutable.List [
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "BOLD",
          ],
          "entity": "1",
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "BOLD",
          ],
          "entity": "1",
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "BOLD",
          ],
          "entity": "1",
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "BOLD",
          ],
          "entity": "1",
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "BOLD",
          ],
          "entity": "1",
        },
      ],
      "depth": 0,
      "data": Immutable.Map {},
    },
    "c": Immutable.Record {
      "key": "c",
      "type": "code-block",
      "text": "Test",
      "characterList": Immutable.List [
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
      ],
      "depth": 0,
      "data": Immutable.Map {},
    },
    "d": Immutable.Record {
      "key": "d",
      "type": "code-block",
      "text": "",
      "characterList": Immutable.List [],
      "depth": 0,
      "data": Immutable.Map {},
    },
    "e": Immutable.Record {
      "key": "e",
      "type": "code-block",
      "text": "",
      "characterList": Immutable.List [],
      "depth": 0,
      "data": Immutable.Map {},
    },
    "f": Immutable.Record {
      "key": "f",
      "type": "blockquote",
      "text": "Charlie",
      "characterList": Immutable.List [
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "ITALIC",
          ],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "ITALIC",
          ],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "ITALIC",
          ],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "ITALIC",
          ],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "ITALIC",
          ],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "ITALIC",
          ],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "ITALIC",
          ],
          "entity": null,
        },
      ],
      "depth": 0,
      "data": Immutable.Map {},
    },
  },
  "selectionBefore": Immutable.Record {
    "anchorKey": "a",
    "anchorOffset": 2,
    "focusKey": "a",
    "focusOffset": 2,
    "isBackward": false,
    "hasFocus": true,
  },
  "selectionAfter": Immutable.Record {
    "anchorKey": "a",
    "anchorOffset": 4,
    "focusKey": "a",
    "focusOffset": 4,
    "isBackward": false,
    "hasFocus": true,
  },
}
`;

exports[`must apply multiblock fragments 1`] = `
Immutable.Record {
  "entityMap": Immutable.OrderedMap {},
  "blockMap": Immutable.OrderedMap {
    "a": Immutable.Record {
      "key": "a",
      "type": "unstyled",
      "text": "xx",
      "characterList": Immutable.List [
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
      ],
      "depth": 0,
      "data": Immutable.Map {
        "a": 1,
      },
    },
    "key0": Immutable.Record {
      "key": "key0",
      "type": "unstyled",
      "text": "yyAlpha",
      "characterList": Immutable.List [
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
      ],
      "depth": 0,
      "data": Immutable.Map {
        "b": 2,
      },
    },
    "b": Immutable.Record {
      "key": "b",
      "type": "unordered-list-item",
      "text": "Bravo",
      "characterList": Immutable.List [
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "BOLD",
          ],
          "entity": "1",
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "BOLD",
          ],
          "entity": "1",
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "BOLD",
          ],
          "entity": "1",
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "BOLD",
          ],
          "entity": "1",
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "BOLD",
          ],
          "entity": "1",
        },
      ],
      "depth": 0,
      "data": Immutable.Map {},
    },
    "c": Immutable.Record {
      "key": "c",
      "type": "code-block",
      "text": "Test",
      "characterList": Immutable.List [
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [],
          "entity": null,
        },
      ],
      "depth": 0,
      "data": Immutable.Map {},
    },
    "d": Immutable.Record {
      "key": "d",
      "type": "code-block",
      "text": "",
      "characterList": Immutable.List [],
      "depth": 0,
      "data": Immutable.Map {},
    },
    "e": Immutable.Record {
      "key": "e",
      "type": "code-block",
      "text": "",
      "characterList": Immutable.List [],
      "depth": 0,
      "data": Immutable.Map {},
    },
    "f": Immutable.Record {
      "key": "f",
      "type": "blockquote",
      "text": "Charlie",
      "characterList": Immutable.List [
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "ITALIC",
          ],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "ITALIC",
          ],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "ITALIC",
          ],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "ITALIC",
          ],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "ITALIC",
          ],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "ITALIC",
          ],
          "entity": null,
        },
        Immutable.Record {
          "style": Immutable.OrderedSet [
            "ITALIC",
          ],
          "entity": null,
        },
      ],
      "depth": 0,
      "data": Immutable.Map {},
    },
  },
  "selectionBefore": Immutable.Record {
    "anchorKey": "a",
    "anchorOffset": 0,
    "focusKey": "a",
    "focusOffset": 0,
    "isBackward": false,
    "hasFocus": true,
  },
  "selectionAfter": Immutable.Record {
    "anchorKey": "key0",
    "anchorOffset": 2,
    "focusKey": "key0",
    "focusOffset": 2,
    "isBackward": false,
    "hasFocus": true,
  },
}
`;
